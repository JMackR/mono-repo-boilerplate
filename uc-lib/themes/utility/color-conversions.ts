// tslint:disable: no-magic-numbers
// tslint:disable: no-bitwise
export const ensureHex = (input: string): string => {
  if (input.startsWith("rgba")) {
    return rgba2hex(input, input.length === 9).toUpperCase()
  } else if (!input.startsWith("#")) {
    // tslint:disable-next-line: no-console
    console.warn("unable to ensure hex", input)
  }
  return input.toUpperCase()
}

/**
 * Adapted from https://stackoverflow.com/questions/49974145/how-to-convert-rgba-to-hex-color-code-using-javascript#49974627
 * @param orig
 * @param forceAlpha
 */
function rgba2hex(orig: string, forceAlpha?: boolean) {
  let a
  const rgb = orig.replace(/\s/g, "").match(/^rgba?\((\d+),(\d+),(\d+),?([^,\s)]+)?/i)
  const alpha = ((rgb && rgb[4]) || "").trim()
  let hex = rgb
    ? (rgb[1] | (1 << 8)).toString(16).slice(1) +
      (rgb[2] | (1 << 8)).toString(16).slice(1) +
      (rgb[3] | (1 << 8)).toString(16).slice(1)
    : orig

  if (alpha !== "") {
    a = alpha
  } else {
    a = 1
  }
  // multiply before convert to HEX
  a = ((a * 255) | (1 << 8)).toString(16).slice(1)
  hex = hex

  if (a !== "ff" || forceAlpha === true) {
    return "#" + a + hex
  } else if (a === "ff") {
    return "#" + hex
  }
  throw Error("unexpected rgba2hex() conversion!")
}
