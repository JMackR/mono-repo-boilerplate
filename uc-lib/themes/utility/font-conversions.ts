import { FontStyle } from "../type-defs"

import { StyleSheet, CSSPropertiesComplete } from "aphrodite/no-important"
/**
 * Converts the FontStyle to an aphrodite StyleSheet object
 * @param font
 */
export const covertFontToWebStyleSheet = (
  font: FontStyle,
  color?: string,
  display: CSSPropertiesComplete["display"] = "inline",
) => {
  const fontStyles = StyleSheet.create({
    text: {
      color,
      fontSize: font.fontSize,
      fontWeight: convertFontWeight(font.fontWeight),
      fontFamily: font.fontFamily + ", sans-serif",
      display,
    },
  })
  return fontStyles
}

/**
 * Convert a font weight for use with aphrodite for web
 * @param weight
 */
// tslint:disable: no-magic-numbers
export const convertFontWeight = (
  weight: string,
):
  | "initial"
  | "inherit"
  | "unset"
  | "normal"
  | "bold"
  | "bolder"
  | "lighter"
  | 100
  | 200
  | 300
  | 400
  | 500
  | 600
  | 700
  | 800
  | 900
  | undefined => {
  switch (weight) {
    case "100":
      return 100
    case "200":
      return 200
    case "300":
      return 300
    case "400":
      return 400
    case "500":
      return 500
    case "600":
      return 600
    case "700":
      return 700
    case "800":
      return 800
    case "900":
      return 900
    case "initial":
      return "initial"
    case "inherit":
      return "inherit"
    case "unset":
      return "unset"
    case "normal":
      return "normal"
    case "bold":
      return "bold"
    case "bolder":
      return "bolder"
    case "lighter":
      return "lighter"
    default:
      return undefined
  }
}

/**
 * Converts an FontTheme line height for use with web.
 * @param lineHeight line height from an FontTheme.
 */
export const convertLineHeightToPx = (lineHeight: number): string => {
  return lineHeight + "px"
}
