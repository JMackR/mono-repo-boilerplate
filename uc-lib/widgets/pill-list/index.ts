export * from "./pill-list"
export * from "./pill"
export * from "./pill-list-props"
