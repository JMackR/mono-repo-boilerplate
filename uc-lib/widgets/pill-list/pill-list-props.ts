import React, { MouseEvent } from "react"
import { NativeTouchEvent } from "react-native"
import { TextTypes, Colors, TextColors } from "uc-lib/themes"

export interface PillDataProps {
  text: string
  textType?: keyof TextTypes
  paddingStepVertical?: number
  paddingStepHorizontal?: number
  icon?: JSX.Element
  href?: string
  pillColor?: keyof Colors
  textColor?: keyof TextColors
  border?: boolean
  borderColor?: keyof Colors
  onClick?(event: NativeTouchEvent | MouseEvent): void
  /**
   * Used to locate this view in end-to-end tests.
   */
  testID?: string
}
export interface PillProps {
  data: PillDataProps
  index: number
}

export interface PillListProps {
  data: PillDataProps[]
  /**
   * Used to locate this view in end-to-end tests.
   */
  paddingLeft?: number
  testID?: string
}
