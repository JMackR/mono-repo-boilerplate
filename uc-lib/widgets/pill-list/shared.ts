export const CONTAINER_MARGIN = 8
export const BORDER_WIDTH = 1

export interface Padding {
  padding?: number
  paddingRight?: number
  paddingLeft?: number
  paddingTop?: number
  paddingBottom?: number
}

const getMargin = (baseMargin: number, marginStep?: number) => {
  if (marginStep) {
    return marginStep * baseMargin
  }
  return baseMargin
}

export const borderWidth = (showBorder?: boolean) => {
  return showBorder ? BORDER_WIDTH : 0
}

export const determinePadding = (
  baseMargin: number,
  paddingStepHorizontal?: number,
  paddingStepVertical?: number,
  border?: boolean,
) => {
  let padding: Padding = {
    padding: CONTAINER_MARGIN - borderWidth(border),
  }

  if (paddingStepHorizontal || paddingStepVertical) {
    const paddingHorizontal =
      (paddingStepHorizontal ? getMargin(baseMargin, paddingStepHorizontal) : CONTAINER_MARGIN) - borderWidth(border)
    const paddingVertical =
      (paddingStepVertical ? getMargin(baseMargin, paddingStepVertical) : CONTAINER_MARGIN) - borderWidth(border)
    padding = {
      paddingRight: paddingHorizontal,
      paddingLeft: paddingHorizontal,
      paddingTop: paddingVertical,
      paddingBottom: paddingVertical,
    }
  }

  return padding
}
