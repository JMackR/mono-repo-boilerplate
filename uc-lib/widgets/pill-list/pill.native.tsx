import React from "react"
import { StyleSheet } from "react-native"
import { Text, Stack, Touchable } from "../../controls"
import { useColor, useMargin } from "../../themes"
import { PillProps } from "./pill-list-props"
import { determinePadding, borderWidth, CONTAINER_MARGIN } from "./shared"

const ICON_MARGIN = 1
const PILL_BORDER_RADIUS = 14

export const Pill: React.FC<PillProps> = ({ data, index }) => {
  const { colors } = useColor()
  const { baseMargin } = useMargin()
  const {
    text,
    icon,
    onClick,
    testID,
    textType,
    paddingStepHorizontal,
    paddingStepVertical,
    pillColor,
    textColor,
    border,
    borderColor,
  } = data
  const padding = determinePadding(baseMargin, paddingStepHorizontal, paddingStepVertical, border)

  const backgroundColor = pillColor ? colors[pillColor!] : colors.limestone

  const STYLES = StyleSheet.create({
    container: {
      backgroundColor,
      borderRadius: PILL_BORDER_RADIUS,
      marginRight: CONTAINER_MARGIN,
      borderColor: border ? (borderColor ? colors[borderColor] : colors.limestone) : undefined,
      borderWidth: borderWidth(border),
      borderStyle: border ? "solid" : undefined,
      ...padding,
    },
  })

  return (
    <Touchable
      onPress={onClick}
      style={STYLES.container}
      disabled={onClick === undefined}
      testID={testID || "uc-lib.pill-list-item" + (index ? "." + index : "")}
    >
      <Stack direction="row" childSeparationStep={ICON_MARGIN}>
        {icon && icon}
        <Text textType={textType ? textType : "secondaryBody1"} color={textColor || "primary"}>
          {text}
        </Text>
      </Stack>
    </Touchable>
  )
}
