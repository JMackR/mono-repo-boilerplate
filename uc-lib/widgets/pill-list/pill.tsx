import React from "react"
import { StyleSheet, css } from "aphrodite/no-important"
import { Text, Stack, Touchable } from "../../controls"
import { useColor, useMargin } from "../../themes"
import { PillProps } from "./pill-list-props"
import { ViewBlock } from "../../controls"
import { determinePadding, borderWidth } from "./shared"

const ICON_MARGIN = 1
const PILL_BORDER_RADIUS = 14
const CONTAINER_MARGIN = 8

export const Pill: React.FC<PillProps> = ({ data, index }) => {
  const { baseMargin } = useMargin()
  const { colors } = useColor()
  const {
    text,
    href,
    icon,
    onClick,
    pillColor,
    textColor,
    testID,
    paddingStepHorizontal,
    paddingStepVertical,
    border,
    borderColor,
  } = data
  const backgroundColor = pillColor ? colors[pillColor!] : colors.limestone
  const padding = determinePadding(baseMargin, paddingStepHorizontal, paddingStepVertical, border)

  const STYLES = StyleSheet.create({
    container: {
      backgroundColor,
      borderRadius: PILL_BORDER_RADIUS,
      marginRight: CONTAINER_MARGIN,
      whiteSpace: "nowrap",
      textDecoration: "none",
      display: "block",
      borderColor: border ? (borderColor ? colors[borderColor] : colors.limestone) : undefined,
      borderWidth: borderWidth(border),
      borderStyle: border ? "solid" : "none",
      ...padding,
    },
  })

  return (
    <Touchable
      href={href}
      onPress={onClick}
      disabled={onClick === undefined}
      testID={testID || "uc-lib.pill-list-item" + (index ? "." + index : "")}
    >
      <ViewBlock className={css(STYLES.container)}>
        <Stack direction="row" childSeparationStep={ICON_MARGIN}>
          {icon && icon}
          <Text textType="secondaryBody2" color={textColor || "primary"}>
            {text}
          </Text>
        </Stack>
      </ViewBlock>
    </Touchable>
  )
}
