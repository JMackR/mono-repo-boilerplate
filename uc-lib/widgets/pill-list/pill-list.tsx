import React from "react"
import { List } from "../../controls/list"
import { Pill } from "./pill"
import { PillDataProps, PillListProps } from "./pill-list-props"

const renderItem = (item: PillDataProps, index: number): JSX.Element => {
  return <Pill data={item} key={index} index={index} />
}

export const PillList: React.FC<PillListProps> = ({ data, paddingLeft, testID }) => {
  return (
    <List
      data={data}
      horizontal={true}
      paddingLeft={paddingLeft}
      renderItem={renderItem}
      testID={testID || "uc-lib.pill-list"}
    />
  )
}
