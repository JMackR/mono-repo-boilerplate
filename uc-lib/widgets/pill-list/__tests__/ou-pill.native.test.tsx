import React from "react"
import { shallow, ShallowWrapper } from "enzyme"
import { render } from "react-native-testing-library"
import { ouMockPlatform, ouResetPlatformMock } from "uc-lib/test/react-native/utility"
import { Pill } from "../pill"
import { PillDataProps } from "../pill-list.d"

describe("<Pill />", () => {
  const createComponent = ({ data }: { data: PillDataProps }) => <Pill data={data} />
  const createWrapper = ({ data }: { data: PillDataProps }) => shallow(<Pill data={data} />)
  let wrapper: ShallowWrapper, data: PillDataProps
  beforeEach(() => {
    data = {
      text: "Test",
      onClick: jest.fn(),
    }
  })
  afterEach(() => {
    ouResetPlatformMock()
  })
  it("should contain relavant component", () => {
    wrapper = createWrapper({ data })
    expect(wrapper.find("Text").props().children).toBe(data.text)
  })
  describe("Snapshot", () => {
    it("Android should render properly ", () => {
      ouMockPlatform("android")
      const tree = render(createComponent({ data }))
      expect(tree).toMatchSnapshot()
    })
    it("iOS should render properly ", () => {
      ouMockPlatform("ios")
      const tree = render(createComponent({ data }))
      expect(tree).toMatchSnapshot()
    })
  })
})
