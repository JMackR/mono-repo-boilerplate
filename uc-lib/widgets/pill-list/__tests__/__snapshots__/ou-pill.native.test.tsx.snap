// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pill /> Snapshot Android should render properly  1`] = `
<View
  accessibilityLabel="redibs-ucl.pill-list-item"
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="redibs-ucl.pill-list-item"
>
  <View
    style={
      Object {
        "backgroundColor": "#E2E2E2",
        "borderColor": undefined,
        "borderRadius": 14,
        "borderStyle": undefined,
        "borderWidth": 0,
        "marginRight": 8,
        "padding": 8,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "stretch",
          "backgroundColor": undefined,
          "flexBasis": "auto",
          "flexDirection": "row",
          "flexGrow": 0,
          "flexShrink": 1,
          "height": undefined,
          "justifyContent": "flex-start",
          "overflow": undefined,
          "width": undefined,
        }
      }
    >
      <Text
        accessibilityLabel="redibs-ucl.text"
        style={
          Array [
            Object {
              "fontFamily": "Mulish-Bold",
              "fontSize": 14,
              "fontWeight": "bold",
              "lineHeight": 16,
            },
            Object {
              "color": "#121212",
              "textAlign": undefined,
              "textDecorationLine": undefined,
            },
          ]
        }
        testID="redibs-ucl.text"
      >
        Test
      </Text>
    </View>
  </View>
</View>
`;

exports[`<Pill /> Snapshot iOS should render properly  1`] = `
<View
  accessibilityLabel="redibs-ucl.pill-list-item"
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="redibs-ucl.pill-list-item"
>
  <View
    style={
      Object {
        "backgroundColor": "#E2E2E2",
        "borderColor": undefined,
        "borderRadius": 14,
        "borderStyle": undefined,
        "borderWidth": 0,
        "marginRight": 8,
        "padding": 8,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "stretch",
          "backgroundColor": undefined,
          "flexBasis": "auto",
          "flexDirection": "row",
          "flexGrow": 0,
          "flexShrink": 1,
          "height": undefined,
          "justifyContent": "flex-start",
          "overflow": undefined,
          "width": undefined,
        }
      }
    >
      <Text
        accessibilityLabel="redibs-ucl.text"
        style={
          Array [
            Object {
              "fontFamily": "Mulish-Bold",
              "fontSize": 14,
              "fontWeight": "bold",
              "lineHeight": 16,
            },
            Object {
              "color": "#121212",
              "textAlign": undefined,
              "textDecorationLine": undefined,
            },
          ]
        }
        testID="redibs-ucl.text"
      >
        Test
      </Text>
    </View>
  </View>
</View>
`;
